psi4/src/psi4/libscf_solver/hf.cc

```diff
@@ -1089,17 +1084,39 @@ void HF::guess() {
         format_guess();
         form_D();
 
-        // This is a guess iteration similar to SAD
+        // This is a guess iteration: orbital occupations may be reset in SCF
         iteration_ = -1;
         guess_E = compute_initial_E();
 
     } else if (guess_type == "SAD") {
         if (print_) outfile->Printf("  SCF Guess: Superposition of Atomic Densities via on-the-fly atomic UHF.\n\n");
 
-        // Superposition of Atomic Density
+        // Superposition of Atomic Density. Modified by Susi Lehtola
+        // 2018-12-15 to work also for ROHF, as well as to allow using
+        // SAD with predefined orbital occupations. The algorithm is
+        // the same as in van Lenthe et al, "Starting SCF Calculations
+        // by Superposition of Atomic Densities", J Comput Chem 27,
+        // 926 (2006).
+
+        // Build SAD density matrix
+        compute_SAD_guess();
+
+        // Form initial spin-restricted Fock matrix from the SAD density
+        form_G();
+        form_initial_F();
+
+        // Diagonalize spin-restricted Fock matrix to get the same
+        // orbitals for alpha and beta, also allowing spin-restricted
+        // open shell calculations to work
+        form_initial_C();
+
+        // Now can form the target density matrix
+        form_D();
```

Comment:
The way this is currently formed it appears that there is now, effectively, an extra SCF iteration for the SAD guess which doesn't seem necessary. I need to think a bit more on the correct steps each guess needs to take, but something is a bit wonky here.

Reply:
Correct, which is why we mark a SAD guess as iteration zero (rather than one) knowing the energy is not variational and letting canonical SCF tech take care of the iteration that you wrote out in the code. Effectively, this code changes the expected Psi4 behavior and iteration 0 will now effectively have two iterations inside of it.

Reply:
SAD is the only guess that is non-variational. A read-in orbital guess gives an idempotent density matrix, if the orbitals are reorthonormalized.

Is the changed behavior a problem, given that the results of the reorganized guess are much better than that of the previous code?

Reply:
Done